dwave/system/composites/reversecomposite.py

```diff
@@ -0,0 +1,205 @@
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# =============================================================================
+"""
+Composites that do batch operations for reverse annealing.
+"""
+
+import dimod
+import numpy as np
+
+__all__ = 'ReverseAdvanceComposite', 'ReverseBatchStatesComposite'
+
+
+class ReverseAdvanceComposite(dimod.ComposedSampler):
+    """Composite that reverse anneals an initial sample through a sequence of anneal
+     schedules. If you don not specify an initial sample, a random sample is used for the first
+     submission. By default, each subsequent submission selects the most-found lowest-energy
+     sample as its initial state. If you set reinitialize_state to False, which makes each submission
+     behave like a random walk, the subsequent submission selects the last returned sample as
+     its initial state.
+
+    Args:
+       sampler (:obj:`dimod.Sampler`):
+            A dimod sampler.
+
+    """
+
+    def __init__(self, child_sampler):
+        self._children = [child_sampler]
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def parameters(self):
+        param = self.child.parameters.copy()
+        param['schedules'] = []
+        return param
+
+    @property
+    def properties(self):
+        return {'child_properties': self.child.properties.copy()}
+
+    def sample(self, bqm, anneal_schedules=None, **parameters):
+        """Sample the binary quadratic model using reverse annealing along a given set of anneal schedules.
+
+        Args:
+            bqm (:obj:`dimod.BinaryQuadraticModel`):
+                Binary quadratic model to be sampled from.
+
+            anneal_schedules (list of lists): Anneal schedules in order of submission. Each schedule is
+                formatted as a list of [time, s] pairs
+
+            **parameters:
+                Parameters for the sampling method, specified by the child sampler.
+
+        Returns:
+            :obj:`dimod.SampleSet` that has initial_state and schedule_index fields.
+
+        """
+        child = self.child
+
+        if anneal_schedules is None:
+            return child.sample(bqm, **parameters)
+
+        vartype_values = list(bqm.vartype.value)
+        if 'initial_state' not in parameters:
+            initial_state = dict(zip(list(bqm.variables), np.random.choice(vartype_values, len(bqm))))
+        else:
+            initial_state = parameters.pop('initial_state')
+
+        if not isinstance(initial_state, dict):
```

Comment:
Maybe use the more abstract `collections.abc.Mapping` which would allow dimod SampleViews etc.

Also, if it handles `initial_state`, that should be called out in the args docs.

Finally I'll just reference https://github.com/dwavesystems/dimod/issues/491 for the future.